File: api/flask_app/app/metrics.py
import os
from posthog import Posthog
from typing import Dict, Optional


class Metrics:
    """Helper to capture metrics for Flask app."""

    def __init__(self) -> None:
        """Should be called during flask app is initialized so env variables are loaded."""
        self.posthog = Posthog(os.environ["POSTHOG_PUBLIC_KEY"],
                               host=os.environ["POSTHOG_PUBLIC_HOST"])

    def capture(self, user_id: str, event_name: str, properties: Optional[Dict] = None):
        """Capture event for given user ID and event name. ID can be anything unique."""
        self.posthog.capture(distinct_id=user_id,
                             event=event_name, properties=properties)

File: extension/src/background/metrics.js
// No external code loading possible (this disables all extensions such as Replay, Surveys, Exceptions etc.)
// Reference: https://posthog.com/docs/libraries/js.
import posthog from "posthog-js/dist/module.no-external";

// Initializes posthog instance if not initialized.
// Problem is that service worker can go idle so it
// needs to be re-initialized.
function initPosthog() {
  if (posthog.__loaded) {
    return;
  }
  posthog.init(process.env.REACT_APP_PUBLIC_POSTHOG_KEY, {
    api_host: process.env.REACT_APP_PUBLIC_POSTHOG_HOST,
    person_profiles: "identified_only",
  });
}

// Identify user with given properties in Posthog.
export function identifyUser(userId, userProperties) {
  initPosthog();
  posthog.identify(userId, userProperties);
}

// Capture event with given properties in Posthog.
export function captureEvent(eventName, properties = null) {
  initPosthog();
  posthog.capture(eventName, properties);
}

// Reset currently identified user in Posthog.
// Usually called upon user logout.
export function resetUserIdentification() {
  initPosthog();
  posthog.reset();
}
